experimental/ast/codec.go

```diff
@@ -0,0 +1,469 @@
+			return true
+		})
+		return &compilerpb.Decl{Decl: &compilerpb.Decl_Body_{Body: proto}}
+
+	case DeclKindRange:
+		decl := decl.AsRange()
+
+		var kind compilerpb.Decl_Range_Kind
+		if decl.IsExtensions() {
+			kind = compilerpb.Decl_Range_KIND_EXTENSIONS
+		} else if decl.IsReserved() {
+			kind = compilerpb.Decl_Range_KIND_RESERVED
+		}
+
+		proto := &compilerpb.Decl_Range{
+			Kind:          kind,
+			Options:       c.options(decl.Options()),
+			Span:          c.span(decl),
+			KeywordSpan:   c.span(decl.Keyword()),
+			SemicolonSpan: c.span(decl.Semicolon()),
+		}
+
+		decl.Iter(func(_ int, e ExprAny) bool {
+			proto.Ranges = append(proto.Ranges, c.expr(e))
+			return true
+		})
+
+		return &compilerpb.Decl{Decl: &compilerpb.Decl_Range_{Range: proto}}
+
+	case DeclKindDef:
+		decl := decl.AsDef()
+
+		var kind compilerpb.Def_Kind
+		switch decl.Classify() {
+		case DefKindMessage:
+			kind = compilerpb.Def_KIND_MESSAGE
+		case DefKindEnum:
+			kind = compilerpb.Def_KIND_ENUM
+		case DefKindService:
+			kind = compilerpb.Def_KIND_SERVICE
+		case DefKindExtend:
+			kind = compilerpb.Def_KIND_EXTEND
+		case DefKindField:
+			kind = compilerpb.Def_KIND_FIELD
+		case DefKindEnumValue:
+			kind = compilerpb.Def_KIND_ENUM_VALUE
+		case DefKindOneof:
+			kind = compilerpb.Def_KIND_ONEOF
+		case DefKindGroup:
+			kind = compilerpb.Def_KIND_GROUP
+		case DefKindMethod:
+			kind = compilerpb.Def_KIND_METHOD
+		case DefKindOption:
+			kind = compilerpb.Def_KIND_OPTION
+		}
+
+		proto := &compilerpb.Def{
+			Kind:          kind,
+			Name:          c.path(decl.Name()),
+			Value:         c.expr(decl.Value()),
+			Options:       c.options(decl.Options()),
+			Span:          c.span(decl),
+			KeywordSpan:   c.span(decl.Keyword()),
+			EqualsSpan:    c.span(decl.Equals()),
+			SemicolonSpan: c.span(decl.Semicolon()),
+		}
+
+		if kind == compilerpb.Def_KIND_FIELD || kind == compilerpb.Def_KIND_UNSPECIFIED {
+			proto.Type = c.type_(decl.Type())
+		}
+
+		if signature := decl.Signature(); !signature.Nil() {
+			proto.Signature = &compilerpb.Def_Signature{
+				Span:        c.span(signature),
+				InputSpan:   c.span(signature.Inputs()),
+				ReturnsSpan: c.span(signature.Returns()),
+				OutputSpan:  c.span(signature.Outputs()),
+			}
+
+			signature.Inputs().Iter(func(_ int, t TypeAny) bool {
+				proto.Signature.Inputs = append(proto.Signature.Inputs, c.type_(t))
+				return true
+			})
+			signature.Outputs().Iter(func(_ int, t TypeAny) bool {
+				proto.Signature.Outputs = append(proto.Signature.Outputs, c.type_(t))
+				return true
+			})
+		}
+
+		if body := decl.Body(); !body.Nil() {
+			proto.Body = &compilerpb.Decl_Body{
+				Span: c.span(decl.Body()),
+			}
+			body.Iter(func(_ int, d DeclAny) bool {
+				proto.Body.Decls = append(proto.Body.Decls, c.decl(d))
+				return true
+			})
+		}
+
+		return &compilerpb.Decl{Decl: &compilerpb.Decl_Def{Def: proto}}
+
+	default:
+		panic(fmt.Sprintf("typeToProto: unknown DeclKind: %d", k))
```

Comment:
Maybe "protoEncoder" as the prefix (here and below)? Or could also just elide the prefix since it's not that clearly helpful (and can obviously get out of sync with its context after refactor/rename operations).

Reply:
I'm just gonna use protocompile/ast, as elsewhere.